aws/rust-runtime/aws-runtime/src/auth.rs

```diff
@@ -178,12 +277,112 @@ pub mod sigv4 {
             let mut builder = SigningParams::builder()
                 .access_key(credentials.access_key_id())
                 .secret_key(credentials.secret_access_key())
-                .region(operation_config.region.as_ref())
-                .service_name(operation_config.service.as_ref())
+                .region(
+                    operation_config
+                        .region
+                        .as_ref()
+                        .ok_or(SigV4SigningError::MissingSigningRegion)?
+                        .as_ref(),
+                )
+                .service_name(
+                    operation_config
+                        .service
+                        .as_ref()
+                        .ok_or(SigV4SigningError::MissingSigningService)?
+                        .as_ref(),
+                )
                 .time(request_timestamp)
                 .settings(settings);
             builder.set_security_token(credentials.session_token());
-            builder.build().expect("all required fields set")
+            Ok(builder.build().expect("all required fields set"))
+        }
+
+        fn extract_operation_config(
+            config_bag: &ConfigBag,
+        ) -> Result<Cow<'_, SigV4OperationSigningConfig>, SigV4SigningError> {
+            let operation_config = config_bag
+                .get::<SigV4OperationSigningConfig>()
+                .ok_or(SigV4SigningError::MissingOperationSigningConfig)?;
+            let endpoint = config_bag
+                .get::<SmithyEndpoint>()
+                .ok_or(SigV4SigningError::NoResolvedEndpoint)?;
+
+            let EndpointAuthSchemeConfig {
+                signing_region_override,
+                signing_service_override,
+            } = Self::extract_endpoint_auth_scheme_config(endpoint)
+                .map_err(SigV4SigningError::BadEndpointAuthSchemeConfig)?;
+
+            match (signing_region_override, signing_service_override) {
+                (None, None) => Ok(Cow::Borrowed(operation_config)),
+                (region, service) => {
+                    let mut operation_config = operation_config.clone();
+                    if region.is_some() {
+                        operation_config.region = region;
+                    }
+                    if service.is_some() {
+                        operation_config.service = service;
+                    }
+                    Ok(Cow::Owned(operation_config))
+                }
+            }
+        }
+
+        fn extract_endpoint_auth_scheme_config(
+            value: &SmithyEndpoint,
+        ) -> Result<EndpointAuthSchemeConfig, EndpointAuthSchemeConfigError> {
```

Comment:
Yeah, not a bad idea. I split the endpoint document parsing so that the orchestrator resolves the selected auth scheme config and passes it directly to the signer.